auth/client/iam/caching.go

```diff
@@ -0,0 +1,198 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package iam
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/nuts-foundation/nuts-node/auth/log"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/pquerna/cachecontrol"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+)
+
+// CachingHTTPRequestDoer is a cache for HTTP responses for DID/OAuth2/OpenID clients.
+// It only caches GET requests (since generally only metadata is cacheable), and only if the response is cacheable.
+// It only works on expiration time and does not respect ETags headers.
+type CachingHTTPRequestDoer struct {
+	maxBytes    int
+	requestDoer core.HTTPRequestDoer
+
+	// currentSizeBytes is the current size of the cache in bytes.
+	// It's used to make room for new entries when the cache is full.
+	currentSizeBytes int
+	// head is the first entry of a linked list of cache entries, ordered by expiration time.
+	// The first entry is the one that will expire first, which optimizes the removal of expired entries.
+	// When an entry is inserted in the cache, it's inserted in the right place in the linked list (ordered by expiry).
+	head *cacheEntry
+	// entriesByURL is a map of cache entries, indexed by the URL of the request.
+	// This optimizes the lookup of cache entries by URL.
+	entriesByURL map[string][]*cacheEntry
+	mux          sync.RWMutex
+}
+
+type cacheEntry struct {
+	responseData    []byte
+	requestURL      *url.URL
+	requestMethod   string
+	requestRawQuery string
+	expirationTime  time.Time
+	next            *cacheEntry
+	responseStatus  int
+	responseHeaders http.Header
+}
+
+func (h *CachingHTTPRequestDoer) Do(httpRequest *http.Request) (*http.Response, error) {
+	if httpRequest.Method == http.MethodGet {
+		if response := h.getCachedEntry(httpRequest); response != nil {
+			return response, nil
+		}
+	}
+
+	httpResponse, err := h.requestDoer.Do(httpRequest)
+	if err != nil {
+		return nil, err
+	}
+	if httpRequest.Method == http.MethodGet {
+		reasons, expirationTime, err := cachecontrol.CachableResponse(httpRequest, httpResponse, cachecontrol.Options{PrivateCache: false})
+		if err != nil {
+			log.Logger().WithError(err).Infof("error while checking cacheability of response (url=%s), not caching", httpRequest.URL.String())
+			return httpResponse, nil
+		}
+		// We don't want to cache responses for too long, as that increases the risk of staleness,
+		// and could keep cause very long-lived entries to never be pruned.
+		maxExpirationTime := time.Now().Add(time.Hour)
+		if expirationTime.After(maxExpirationTime) {
+			expirationTime = maxExpirationTime
+		}
+		if len(reasons) > 0 || expirationTime.IsZero() {
+			log.Logger().Debugf("response (url=%s) is not cacheable: %v", httpRequest.URL.String(), reasons)
+			return httpResponse, nil
+		}
+		responseBytes, err := io.ReadAll(httpResponse.Body)
+		if err != nil {
+			return nil, fmt.Errorf("error while reading response body for caching: %w", err)
+		}
+		h.mux.Lock()
```

Comment:
For readability purposes place a single lock/unlock at the beginning of the only public method.

Reply:
That causes IO to be included in the lock scope, so that causes scalability/performance issues. I refactored the function a bit to make it seem less random.